File: Deepspace-Ruby/lib/SpaceCity.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'SpaceStation'
module Deepspace
class SpaceCity<SpaceStation
 
  def initialize(base, rest)
    super(base.name, SuppliesPackage.new(base.ammoPower, base.fuelUnits, base.shieldPower))
    @base=base
    setLoot(Loot.new(0,0,0,0,base.nMedals, false, false))   #Hacemos base.medals para que se guarden las medallas que había en la estación espacial
    
    h=Hangar.new(1)
    receiveHangar(h)
    for i in 0...base.weapons.size
      receiveWeapon(base.weapons.at(i))
      
      mountWeapon(0)
    end
    
    for i in 0...base.shieldBoosters.size
      receiveShieldBooster(base.shieldBoosters.at(i))
      indice=base.hangar.shieldBoosters.size-1
      mountShieldBooster(indice)
    end
    discardHangar
    receiveHangar(base.hangar)
    setPendingDamage(base.pendingDamage)
    @collaborators=rest
  end
  
  def getCollaborators
    @collaborators
  end
  
  def fire
    fire=@base.fire
    if (@collaborators!=nil)
    for i in 0...@collaborators.size
      fire+=@collaborators[i].fire
    end
    end
    fire
  end
  
  def protection 
    protection=@base.protection
    if (@collaborators!=nil)
      for i in 0 ... @collaborators.size
        protection+=@collaborators[i].protection
      end
    end
  end
  
  def setLoot(loot)
    super(Loot.new(loot.nSupplies, loot.nWeapons, loot.nShields,  loot.nHangars, loot.nMedals, false, false));
  end
  
  def to_s
    salida="SpaceCity={"
    salida+="Base= "+super
    if !@collaborators.nil?
      for i in 0...@collaborators.size
        salida+="       Colaborador "+i.to_s+ "{ "+@collaborators[i].to_s + "}\n"
      end
    end
    salida
    
  end
end
end

File: Deepspace-Java/src/deepspace/SpaceCity.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author inaki
 */
public class SpaceCity extends SpaceStation{
    ArrayList<SpaceStation> collaborators;
    SpaceStation base;
    
    public SpaceCity(SpaceStation base, ArrayList<SpaceStation> rest){
        super(base.getName(),new SuppliesPackage(base.getAmmoPower(), base.getFuelUnits(), base.getShieldPower()) );
        this.base=base;
        setLoot(new Loot(0,0,0,0,base.getNMedals(), false, false));
        
        setPendingDamage(base.getPendingDamage());
        this.collaborators=new ArrayList(rest);
                Hangar h=new Hangar(1);
        receiveHangar(h);
        
        if (base.getWeapons()!=null){
            for (int i=0; i< base.getWeapons().size();i++){
                  super.receiveWeapon(base.getWeapons().get(i));
                  super.mountWeapon(base.getHangar().getWeapons().size()-1);
            }
        }
        if (base.getShieldBoosters()!=null){
            for (int i=0; i<base.getShieldBoosters().size(); i++){
                super.receiveShieldBooster(base.getShieldBoosters().get(i));
                super.mountShieldBooster(base.getHangar().getShieldBoosters().size()-1);
            }
        }
        super.discardHangar();
        super.receiveHangar(base.getHangar());
        
        
    }
    
    public ArrayList<SpaceStation> getCollaborators(){
        return this.collaborators;
    }
    
    @Override
    public float fire(){
        float firetotal=0;
        firetotal+=base.fire();
        for (int i=0; i < collaborators.size(); i++)
            firetotal+=collaborators.get(i).fire();
        return firetotal;       
    }
    
    @Override       //Redefinirlo: Mismos parametros, sobreescribirlos: distintos 
    public float protection(){
        float protectiontotal=0;
        protectiontotal+=base.protection();
        for (int i=0; i< collaborators.size(); i++)
            protectiontotal+=collaborators.get(i).protection();
        
        return protectiontotal;
    }
    
    @Override
    public Transformation setLoot(Loot loot){
        Transformation t=super.setLoot(new Loot(loot.getNSupplies(), loot.getNWeapons(), loot.getNShields(), loot.getNHangars(), loot.getNMedals(), false, false));
        return t;
    }
    
    @Override 
    public String toString(){
        String salida="SpaceCity={ Base= ";
        salida+=base;
        salida+=collaborators;
        return salida;
    }
           
    
    
}
